portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -29,46 +29,64 @@
+    private static final String timestamp = "timestamp";
+    private static final String COMMENT_PARAMETER = "Comment";
+    private static final int defaultMaxComments = 10;
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+      Gson gson = new Gson();
+      int limit = getRequestNum(request);
+      ArrayList<Comment> comments = getComments(limit);
 
-        Comment comment = new Comment(content, timestamp);
-        comments.add(comment);
+      response.setContentType("application/json;");
+      response.getWriter().println(gson.toJson(comments));
     }
-    Gson gson = new Gson();
 
-    response.setContentType("application/json;");
-    response.getWriter().println(gson.toJson(comments));
-  }
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String text = request.getParameter("comment-input");
+        long time = System.currentTimeMillis();
 
-  @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String content = request.getParameter("comment-input");
-    long timestamp = System.currentTimeMillis();
+        Entity comment = new Entity("Comment");
+        comment.setProperty(content, text);
+        comment.setProperty(timestamp,time);
 
-    Entity comment = new Entity(COMMENT_PARAMETER);
-    comment.setProperty(CONTENT_PARAMETER, content);
-    comment.setProperty(TIMESTAMP_PARAMETER, timestamp);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(comment);
+        int max = getRequestNum(request);  
+        response.sendRedirect("/index.html?limit=" + max);
+    }
+
+    private int getRequestNum(HttpServletRequest request) {
+        String limitString = request.getParameter("limit");
+        int limit;
+        try {
+            limit = Integer.parseInt(limitString);
+        } catch(NumberFormatException e) {
+            return defaultMaxComments;
+        }
+        return limit;
+    } 
 
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    datastore.put(comment);
+    private ArrayList<Comment> getComments(int limit) {
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery res = datastore.prepare(query);
 
-    response.sendRedirect("/index.html");
-  }
-}
+        ArrayList<Comment> comments = new ArrayList<Comment>();
+        for(Entity entity : res.asIterable()) {
+            String content = (String) entity.getProperty("content");
+            long timestamp = (long) entity.getProperty("timestamp");
+            Comment comment = new Comment(content, timestamp);
+            comments.add(comment);
+            if(comments.size() > limit) {
```

Comment:
Since this is strict greater-than instead of greater-than-or-equal than, I'm seeing that you're getting one more comment than requested e.g. when I ask for 3 comments, I actually get 4.